core/block_dragger.js

```diff
@@ -0,0 +1,239 @@
+  this.draggedConnectionManager_.update(currentDragDeltaXY, this.deleteArea_);
+  this.updateCursorDuringBlockDrag_();
+};
+
+/**
+ * Finish a block drag and put the block back on the workspace.
+ * @param {!Event} e The mouseup/touchend event.
+ * @param {!goog.math.Coordinate} currentDragDeltaXY How far the pointer has
+ *     moved from the position at the start of the drag.
+ */
+Blockly.BlockDragger.prototype.endBlockDrag = function(e, currentDragDeltaXY) {
+  // Make sure internal state is fresh.
+  this.dragBlock(e, currentDragDeltaXY);
+  Blockly.BlockSvg.disconnectUiStop_();
+  // TODO: Consider where moveOffDragSurface should live.
+  this.draggingBlock_.moveConnections_(currentDragDeltaXY.x,
+      currentDragDeltaXY.y);
+  this.draggingBlock_.moveOffDragSurface_();
+  var deleted = this.maybeDeleteBlock_();
+  if (!deleted) {
+    // setDragging_ is expensive and doesn't need to be done if we're deleting.
+    this.draggingBlock_.setDragging_(false);
+    this.draggedConnectionManager_.applyConnections();
+    this.draggingBlock_.render();
+    this.fireMoveEvent_();
+  }
+  this.workspace_.setResizesEnabled(true);
+  Blockly.Css.setCursor(Blockly.Css.Cursor.OPEN);
+};
+
+/**
+ * Fire a move event at the end of a block drag.
+ * @private
+ */
+Blockly.BlockDragger.prototype.fireMoveEvent_ = function() {
+  var event = new Blockly.Events.Move(this.draggingBlock_);
+  event.oldCoordinate = this.startCoordinate_;
+  event.recordNew();
+  Blockly.Events.fire(event);
+  var draggingBlock = this.draggingBlock_;
+  // Ensure that any snap and bump are part of this move's event group.
+  var group = Blockly.Events.getGroup();
+  setTimeout(function() {
+    Blockly.Events.setGroup(group);
+    draggingBlock.snapToGrid();
+    Blockly.Events.setGroup(false);
+  }, Blockly.BUMP_DELAY / 2);
+  setTimeout(function() {
+    Blockly.Events.setGroup(group);
+    draggingBlock.bumpNeighbours_();
+    Blockly.Events.setGroup(false);
+  }, Blockly.BUMP_DELAY);
+};
+
+/**
+ * Shut the trash can and, if necessary, delete the dragging block.
+ * Should be called at the end of a block drag.
+ * @return {boolean} whether the block was deleted.
+ * @private
+ */
+Blockly.BlockDragger.prototype.maybeDeleteBlock_ = function() {
+  // TODO: is there an event to fire?
+  var trashcan = this.workspace_.trashcan;
+
+  if (this.wouldDeleteBlock_) {
+    if (trashcan) {
+      goog.Timer.callOnce(trashcan.close, 100, trashcan);
+    }
+    this.draggingBlock_.dispose(false, true);
+  } else if (trashcan) {
+    // Make sure the trash can is closed.
+    trashcan.close();
+  }
+  return this.wouldDeleteBlock_;
+};
+
+/**
+ * Update the cursor (and possibly the trash can lid) to reflect whether the
+ * dragging block would be deleted if released immediately.
+ * @private
+ */
+Blockly.BlockDragger.prototype.updateCursorDuringBlockDrag_ = function() {
+  this.wouldDeleteBlock_ = this.draggedConnectionManager_.wouldDeleteBlock();
```

Comment:
I think you should pull this logic up into this class. The ConnectionManager doesn't need to know about the move event, does it?